File: packages/scout/src/cherrypick/scout/analytics/levels.py
"""SMA overlays and swing-extrema support/resistance clustering, computed purely from a list of daily
bars (``{"t","o","h","l","c","v"}``, oldest first) -- no I/O. Import-clean (stdlib only) so a future
promotion to ``cherrypick.core.levels`` is a file move once stable, same posture as ``payoff``/``pop``
(see the package README).
"""

from __future__ import annotations

SMA_WINDOWS = (20, 50, 200)


def sma(closes: list[float], window: int) -> list[float | None]:
    """Simple moving average, same length as `closes`. The first `window - 1` entries are `None`
    (not enough history yet) rather than an average over a short window pretending to be the real
    thing."""
    out: list[float | None] = [None] * len(closes)
    running = 0.0
    for i, c in enumerate(closes):
        running += c
        if i >= window:
            running -= closes[i - window]
        if i >= window - 1:
            out[i] = running / window
    return out


def moving_averages(bars: list[dict]) -> dict[str, list[float | None]]:
    """`{"sma20": [...], "sma50": [...], "sma200": [...]}`, each aligned to `bars` by index."""
    closes = [b["c"] for b in bars]
    return {f"sma{window}": sma(closes, window) for window in SMA_WINDOWS}


def _swing_extrema(bars: list[dict], *, lookback: int) -> tuple[list[float], list[float]]:
    """A bar is a swing high/low when its high/low is the *unique* extremum of the `lookback` bars on
    both sides of it. Uniqueness matters: a plain max/min check fires on every bar of a flat run (each
    one ties the window extremum), which would report a swing at every point of a sideways market --
    not the turn a swing is supposed to mark."""
    highs = [b["h"] for b in bars]
    lows = [b["l"] for b in bars]
    n = len(bars)
    swing_highs, swing_lows = [], []
    for i in range(lookback, n - lookback):
        window = slice(i - lookback, i + lookback + 1)
        h_window, l_window = highs[window], lows[window]
        if highs[i] == max(h_window) and h_window.count(highs[i]) == 1:
            swing_highs.append(highs[i])
        if lows[i] == min(l_window) and l_window.count(lows[i]) == 1:
            swing_lows.append(lows[i])
    return swing_highs, swing_lows


def _cluster(prices: list[float], tolerance_pct: float) -> list[tuple[float, int]]:
    """Merge swing prices within `tolerance_pct` of their neighbor into one level (mean price, touch
    count) -- several swings pinning the same zone should read as one level, not a cluttered many."""
    if not prices:
        return []
    ordered = sorted(prices)
    groups: list[list[float]] = [[ordered[0]]]
    for price in ordered[1:]:
        if abs(price - groups[-1][-1]) / groups[-1][-1] <= tolerance_pct:
            groups[-1].append(price)
        else:
            groups.append([price])
    return [(sum(g) / len(g), len(g)) for g in groups]


def support_resistance(bars: list[dict], *, lookback: int = 3, tolerance_pct: float = 0.005) -> list[dict]:
    """`[{"price", "kind": "support"|"resistance", "touches"}, ...]`, sorted by price."""
    swing_highs, swing_lows = _swing_extrema(bars, lookback=lookback)
    levels = [
        {"price": price, "kind": "resistance", "touches": touches}
        for price, touches in _cluster(swing_highs, tolerance_pct)
    ] + [
        {"price": price, "kind": "support", "touches": touches}
        for price, touches in _cluster(swing_lows, tolerance_pct)
    ]
    levels.sort(key=lambda level: level["price"])
    return levels

File: packages/console/server/src/analytics/levels.ts
/**
 * Port of scout's analytics/levels.py: SMA overlays and swing-extrema
 * support/resistance clustering, computed purely from daily bars
 * ({t,o,h,l,c,v}, oldest first). Pure functions, no I/O.
 */

export interface Bar {
  t: number;
  o: number;
  h: number;
  l: number;
  c: number;
  v: number;
}

export interface Level {
  price: number;
  kind: "support" | "resistance";
  touches: number;
}

export const SMA_WINDOWS = [20, 50, 200] as const;

/** SMA aligned to closes; the first window-1 entries are null (not enough history). */
export function sma(closes: number[], window: number): Array<number | null> {
  const out: Array<number | null> = new Array<number | null>(closes.length).fill(null);
  let running = 0;
  for (let i = 0; i < closes.length; i++) {
    running += closes[i]!;
    if (i >= window) running -= closes[i - window]!;
    if (i >= window - 1) out[i] = running / window;
  }
  return out;
}

export function movingAverages(bars: Bar[]): Record<string, Array<number | null>> {
  const closes = bars.map((b) => b.c);
  const out: Record<string, Array<number | null>> = {};
  for (const w of SMA_WINDOWS) out[`sma${w}`] = sma(closes, w);
  return out;
}

/**
 * A bar is a swing high/low when its high/low is the UNIQUE extremum of the
 * lookback bars on both sides — uniqueness so a flat run doesn't report a
 * swing at every bar of a sideways market.
 */
function swingExtrema(bars: Bar[], lookback: number): { highs: number[]; lows: number[] } {
  const highs = bars.map((b) => b.h);
  const lows = bars.map((b) => b.l);
  const swingHighs: number[] = [];
  const swingLows: number[] = [];
  for (let i = lookback; i < bars.length - lookback; i++) {
    const hWindow = highs.slice(i - lookback, i + lookback + 1);
    const lWindow = lows.slice(i - lookback, i + lookback + 1);
    const h = highs[i]!;
    const l = lows[i]!;
    if (h === Math.max(...hWindow) && hWindow.filter((x) => x === h).length === 1) swingHighs.push(h);
    if (l === Math.min(...lWindow) && lWindow.filter((x) => x === l).length === 1) swingLows.push(l);
  }
  return { highs: swingHighs, lows: swingLows };
}

/** Merge swing prices within tolerance of their neighbor into one (mean, touches) level. */
function cluster(prices: number[], tolerancePct: number): Array<[number, number]> {
  if (prices.length === 0) return [];
  const ordered = [...prices].sort((a, b) => a - b);
  const groups: number[][] = [[ordered[0]!]];
  for (const price of ordered.slice(1)) {
    const lastGroup = groups[groups.length - 1]!;
    const lastPrice = lastGroup[lastGroup.length - 1]!;
    if (Math.abs(price - lastPrice) / lastPrice <= tolerancePct) lastGroup.push(price);
    else groups.push([price]);
  }
  return groups.map((g) => [g.reduce((a, b) => a + b, 0) / g.length, g.length]);
}

export function supportResistance(bars: Bar[], lookback = 3, tolerancePct = 0.005): Level[] {
  const { highs, lows } = swingExtrema(bars, lookback);
  const levels: Level[] = [
    ...cluster(highs, tolerancePct).map(([price, touches]): Level => ({ price, kind: "resistance", touches })),
    ...cluster(lows, tolerancePct).map(([price, touches]): Level => ({ price, kind: "support", touches })),
  ];
  levels.sort((a, b) => a.price - b.price);
  return levels;
}
